File: apps/api/app/schemas/gsc.py
"""Schemas for GSC integration endpoints."""

import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel


class GSCAuthUrlResponse(BaseModel):
    auth_url: str
    state: str


class GSCConnectRequest(BaseModel):
    client_id: uuid.UUID
    gsc_property: str


class GSCPropertyItem(BaseModel):
    site_url: str
    permission_level: str


class GSCConnectionResponse(BaseModel):
    id: uuid.UUID
    client_id: uuid.UUID
    gsc_property: str
    connected_by: str
    last_fetch_at: datetime | None
    created_at: datetime

    model_config = {"from_attributes": True}


class GSCKeywordRow(BaseModel):
    query: str
    clicks: int
    impressions: int
    ctr: float
    position: float


class GSCPageRow(BaseModel):
    page: str
    clicks: int
    impressions: int
    ctr: float
    position: float


class GSCSnapshotResponse(BaseModel):
    date_range_start: str
    date_range_end: str
    top_queries: list[dict[str, Any]]
    top_pages: list[dict[str, Any]]
    opportunity_queries: list[dict[str, Any]]
    device_breakdown: dict[str, Any]
    country_breakdown: list[dict[str, Any]]
    fetched_at: datetime

    model_config = {"from_attributes": True}

File: apps/web/src/types/gsc.ts
export interface GSCConnectionStatus {
  connected: boolean;
  gsc_property?: string;
  last_fetch_at?: string;
  created_at?: string;
}

export interface GSCProperty {
  site_url: string;
  permission_level: string;
}

export interface GSCKeywordRow {
  query: string;
  clicks: number;
  impressions: number;
  ctr: number;
  position: number;
}

export interface GSCPageRow {
  page: string;
  clicks: number;
  impressions: number;
  ctr: number;
  position: number;
}

export interface GSCSnapshot {
  date_range_start: string;
  date_range_end: string;
  top_queries: GSCKeywordRow[];
  top_pages: GSCPageRow[];
  opportunity_queries: GSCKeywordRow[];
  device_breakdown: Record<string, { clicks: number; impressions: number }>;
  country_breakdown: Array<{ country: string; clicks: number; impressions: number }>;
  fetched_at: string;
}
